document/core/exec/instructions.rst

```diff
@@ -666,33 +677,51 @@ Memory Instructions
 :math:`\MEMORYGROW`
 ...................
 
-.. todo:: update text to match formalism
+1. Assert: due to :ref:`validation <valid-memory.grow>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+2. Pop the value :math:`\I32.\CONST~n` from the stack.
 
-2. Assert: due to :ref:`validation <valid-memory.grow>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+3. Let :math:`\X{err}` be the |i32| value :math:`2^{32}-1`, for which :math:`\signed_{32}(\X{err})` is :math:`-1`.
 
-3. Let :math:`a` be the :ref:`memory address <syntax-memaddr>` :math:`F.\AMODULE.\MIMEMS[0]`.
+4. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+
+5. Assert: due to :ref:`validation <valid-memory.grow>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+
+6. Let :math:`a` be the :ref:`memory address <syntax-memaddr>` :math:`F.\AMODULE.\MIMEMS[0]`.
+
+7. If the memory is local, i.e., :math:`S.\SMEMS[a]` exists, then:
+
+   a. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
+
+   b. Let :math:`\X{sz}` be the length of :math:`S.\SMEMS[a]` divided by the :ref:`page size <page-size>`.
+
+   c. Either, try :ref:`growing <grow-mem>` :math:`\X{mem}` by :math:`n` :ref:`pages <page-size>`:
+
+      i. If it succeeds, push the value :math:`\I32.\CONST~\X{sz}` to the stack.
+
+      ii. Else, push the value :math:`\I32.\CONST~\X{err}` to the stack.
+
+   d. Or, push the value :math:`\I32.\CONST~\X{err}` to the stack.
 
-4. Assert: due to :ref:`validation <valid-memory.grow>`, :math:`S.\SMEMS[a]` exists.
+8. Else:
 
-5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
+   a. Either:
```

Comment:
the "try growing memory" text in the non-shared case above was always informal language. Could some of that flavour be added to this case too? i.e.

```
a. Either:
  memory growth succeeds.
  ... let bla bla bla
b. Or:
  memory growth fails.
  ... bla bla bla
```